### PR TITLE
Fix ExecSync support for runtimes other than runC

### LINKED ISSUE #309: ExecSync code path does not work with all OCI compatible runtimes
The `ExecSync` code path and tests expects to be run with `runc`:

- It is only waiting for the runtime pid to get the command exit status (and does not support waiting from any other process created by the runtime).
- Negative testing are comparing the error strings with `runc` specific output.

As a consequence CRI-O `ExecSync` fails on e.g. Clear Containers.

### PR DESCRIPTION
**Pull request by sameo**
Some OCI container runtimes (in particular the hypervisor based ones like e.g. Clear Containers) will typically create a shim process between the hypervisor and the runtime caller, in order to not rely on the hypervisor process for e.g. forwarding the output streams or getting a command exit code.

When executing a command inside a running container those runtimes will create that shim process and terminate.
Therefore calling and monitoring the runtime process directly from ExecSync() will fail. Instead we need to have a subreaper calling the runtime and monitoring the shim process, and conmon seems to be a natural place for doing so.

This PR does mostly 2 things:

1) Add a `-e` option to `conmon` to also monitor potential shim processes created by `<runtime> exec`
2) Modify `ExecSync` to call `conmon` instead of calling the runtime directly. `ExecSync` also now reads the exit status code back through the synchronization pipe.

Fixes #309

### CODE REVIEW

File: test/ctr.bats [hunk 1]
```diff
@@ -387,7 +387,6 @@ function teardown() {
 	run ocic ctr execsync --id "$ctr_id" doesnotexist
 	echo "$output"
 	[ "$status" -ne 0 ]
-	[[ "$output" =~ "executable file not found in" ]]
```
Comment (runcom) [inline test/ctr.bats]: Where has this gone?

File: oci/oci.go [hunk 2]
```diff
@@ -179,14 +184,51 @@ func (e ExecSyncError) Error() string {
 
 // ExecSync execs a command in a container and returns it's stdout, stderr and return code.
 func (r *Runtime) ExecSync(c *Container, command []string, timeout int64) (resp *ExecSyncResponse, err error) {
-	args := []string{"exec", c.name}
+	parentPipe, childPipe, err := os.Pipe()
```
Comment (runcom) [inline oci/oci.go]: I'd create a function in the `oci` package to wrap all of this? What do you think?

**Comment by runcom** [comment]
I'm overall ok to have `conmon exec`, @mrunalp @cyphar PTAL

File: test/ctr.bats (hunk 1 above)
Comment (sameo) [inline test/ctr.bats]: This string was returned by runC itself. Other runtimes may not decide to return the same string.
I think the only thing we can check here is that this execsync really failed.

File: oci/oci.go (hunk 2 above)
Comment (sameo) [inline oci/oci.go]: By "all of this" do you mean the pipe and temporary file creation ?

Comment (runcom) [inline oci/oci.go]: > By "all of this" do you mean the pipe and temporary file creation ?

yes

**Comment by laijs** [comment]
> When executing a command inside a running container those runtimes will create that shim process and terminate.

I don't understand it, if the argument "-d" is not added to "runtime exec", the runtime exec process must wait the new process of the container.

**Comment by sameo** [comment]
@laijs You're correct. I'll try playing with that. It does not work currently with clear containers, but it may be a simpler path.

**Comment by laijs** [comment]
I like this approach also, but I think you need to add "-d" to "runtime exec" in ExecSync()

**Comment by sameo** [comment]
@laijs Done, conmon now calls exec with -d.

**Comment by mrunalp** [comment]
Thanks! I will test this out today.

File: conmon/conmon.c [hunk 3]
```diff
@@ -303,25 +323,52 @@ int main(int argc, char *argv[])
 
 	/* Wait for the container process and record its exit code */
 	while ((pid = waitpid(-1, &status, 0)) > 0) {
-		printf("PID %d exited\n", pid);
+		int exit_status = WEXITSTATUS(status);
```
Comment (mrunalp) [inline conmon/conmon.c]: We were doing this conversion up in the go code so need to remove the conversion from https://github.com/kubernetes-incubator/cri-o/blob/master/oci/oci.go#L338

Comment (sameo) [inline conmon/conmon.c]: Done, thanks for spotting that one.

**Comment by mrunalp** [comment]
LGTM

**Comment by cyphar** [comment]
Sorry that I was on holiday while this patch was being reviewed. I don't like this because now all of `conmon` has bifoccated logic for `exec` and non-`exec`. The code is getting really hard to follow...

**Comment by cyphar** [comment]
I'll do a bit of cleanup in #162 ...

**Comment by sameo** [comment]
@cyphar Sounds good. Please ping me when you have it ready so that I can verify it does not break Clear Containers.
